### PR TITLE
Backend for sending notificatons to users

### PR DESCRIPTION
**Pull request by OluwadaraO**
Implementation of notifications in my backend file.

### CODE REVIEW

**Comment by OluwadaraO** [comment]
Users can allow or enable notifications from Culinary Canvas. 
This is a demo of my website: https://www.loom.com/share/2d815cb046484a59b2afb414417a0924

No notifications were shown because notifications are sent at 12am.

File: culinary-canvas/capstone-project/frontend/my_capstone_project/src/UserProfile.jsx [hunk 1]
```diff
@@ -77,6 +129,22 @@ function UserProfile() {
...
....441-67.559-158.622v-84.578c0-81.402-49.742-151.399-120.427-181.203     C381.969,34,347.883,0,306.001,0c-41.883,0-75.968,34.002-76.121,75.849c-70.682,29.804-120.425,99.801-120.425,181.203v84.578     c-0.046,3.181-2.522,129.251-67.561,158.622c-7.409,3.347-11.481,11.412-9.768,19.36c1.711,7.949,8.74,13.626,16.871,13.626     h164.88c3.38,18.594,12.172,35.892,25.619,49.903c17.86,18.608,41.479,28.856,66.502,28.856     c25.025,0,48.644-10.248,66.502-28.856c13.449-14.012,22.241-31.311,25.619-49.903h164.88c8.131,0,15.159-5.676,16.872-13.626     C581.586,511.664,577.516,503.6,570.107,500.254z M484.434,439.859c6.837,20.728,16.518,41.544,30.246,58.866H97.32     c13.726-17.32,23.407-38.135,30.244-58.866H484.434z M306.001,34.515c18.945,0,34.963,12.73,39.975,30.082     c-12.912-2.678-26.282-4.09-39.975-4.09s-27.063,1.411-39.975,4.09C271.039,47.246,287.057,34.515,306.001,34.515z      M143.97,341.736v-84.685c0-89.343,72.686-162.029,162.031-162.029s162.031,72.686,162.031,162.029v84.826     c0.023,2.596,0.427,29.879,7.303,63.465H136.663C143.543,371.724,143.949,344.393,143.97,341.736z M306.001,577.485     c-26.341,0-49.33-18.992-56.709-44.246h113.416C355.329,558.493,332.344,577.485,306.001,577.485z"/>
```
Comment (ayoubmansaratwork) [inline culinary-canvas/capstone-project/frontend/my_capstone_project/src/UserProfile.jsx:137]: Can you please help me understand what this is?

Is this an image? If so, can we just link to the image itself (e.g. `.png`, `.jpg`, etc)

File: culinary-canvas/capstone-project/frontend/my_capstone_project/src/useNotification.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                        reject(new Error('service worker became redundant'))
+                    }
+                });
+            }else{
+                reject(new Error('No service worker found in registration'))
+            }
+        })
+    }
+
+    function urlBase64ToUint8Array(base64String){
```
Comment (ayoubmansaratwork) [inline culinary-canvas/capstone-project/frontend/my_capstone_project/src/useNotification.js:61]: Why is this needed?

File: culinary-canvas/capstone-project/frontend/my_capstone_project/src/useNotification.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                        }),
+                        credentials: 'include'
+                    });
+                    if(!response.ok){
+                        throw new Error('Failed to subscribe')
+                       }
+                    setIsSubscribed(true)
+                }catch(error){
+                console.error('Failed to subscribe to notifications: ', error)
+                throw error;}
```
Comment (ayoubmansaratwork) [inline culinary-canvas/capstone-project/frontend/my_capstone_project/src/useNotification.js:110]: nit: formatting

File: culinary-canvas/capstone-project/backend/prisma/schema.prisma [hunk 4]
```diff
@@ -25,6 +25,44 @@ model Users{
...
+  read Boolean @default(false)
+  createdAt DateTime @default(now())
+  user Users @relation(fields: [userId], references:[id], onDelete: Cascade)
+}
+
+model NotificationSubscription{
+  id Int @id @default(autoincrement())
+  userId Int @unique
+  endpoint String
+  p256dh String
```
Comment (ayoubmansaratwork) [inline culinary-canvas/capstone-project/backend/prisma/schema.prisma:48]: What is this variable?

**Review by ayoubmansaratwork — COMMENTED** [review]
Some questions —

1. Can you update the test plan to show the notifications working? I know you said they are sent out at 12am, but if the user opens the site in the morning (e.g. at 9am) they should be able to see the notifications, right?
2. Is it not possible to send the notification in real-time? Sending it out only at 12am means a potential maximum 24h delay until a user will receive the next notification.

File: culinary-canvas/capstone-project/frontend/my_capstone_project/src/UserProfile.jsx (hunk 1 above)
Comment (OluwadaraO) [reply culinary-canvas/capstone-project/frontend/my_capstone_project/src/UserProfile.jsx:137]: I found it on a site and they wouldn't give the free png so I had to get the svg file. I would find a png file.

File: culinary-canvas/capstone-project/frontend/my_capstone_project/src/useNotification.js (hunk 2 above)
Comment (OluwadaraO) [reply culinary-canvas/capstone-project/frontend/my_capstone_project/src/useNotification.js:61]: When I was checking online, push notifications need the public vapid key in base 8 but the public vapid key is in base 16 so I need the function to convert it so it's able to work as needed.

File: culinary-canvas/capstone-project/backend/prisma/schema.prisma (hunk 4 above)
Comment (OluwadaraO) [reply culinary-canvas/capstone-project/backend/prisma/schema.prisma:48]: So for the subscriptions every user has a p256dh and auth. These keys are used for encryption and servers use it to encrypt and decode the subscription sent.

File: culinary-canvas/capstone-project/frontend/my_capstone_project/src/NotificationModal.jsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    };
+    return(
+        <div className='modal-overlay'>
+            <div className='modal-content'>
+                <h3>Stay Updated With Culinary Canvas</h3>
+                <p>Enable notifications to get the lastest dietary tips!</p>
+                <button onClick={handleSubscribe} disabled={isSubscribed}>
+                    {isSubscribed ? 'Notifications Enabled' : 'Enable Notifications'}
+                </button>
+                <button onClick={onClose}>Maybe Later</button>
```
Comment (joelmet-meta) [inline culinary-canvas/capstone-project/frontend/my_capstone_project/src/NotificationModal.jsx]: if the user is subscribed why show either of these buttons? Wouldn't it just be kinda annoying?

File: culinary-canvas/capstone-project/frontend/my_capstone_project/src/UserProfile.jsx (hunk 1 above)
Comment (joelmet-meta) [reply culinary-canvas/capstone-project/frontend/my_capstone_project/src/UserProfile.jsx:137]: that's too bad :(

File: culinary-canvas/capstone-project/frontend/my_capstone_project/src/Notification.jsx [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+                        method: 'GET',
+                        credentials: 'include',
+                    });
+                    const data = await response.json();
+                    setNotifications(data)
+                }catch(error){
+                    console.error('Error fetching notifications: ', error)
+                }finally{
+                    setLoading(false);
+                }
```
Comment (joelmet-meta) [inline culinary-canvas/capstone-project/frontend/my_capstone_project/src/Notification.jsx]: nit: little funky that there's no space between try/catch/finally and the brackets before/after

File: culinary-canvas/capstone-project/frontend/my_capstone_project/src/Notification.jsx [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                <button>Back to User's Page</button>
+            </Link>
+
+            {notifications.length > 0  && user ? (
+            <ul>
+            {notifications.map((notification) => (
+                <li key={notification.id} style={{backgroundColor: notification.read ? 'lightgrey' : 'white'}}>
+                    <p>{notification.message}</p>
+                    <button onClick={() => deleteNotification(notification.id)}>Delete</button>
+                    {!notification.read &&
```
Comment (joelmet-meta) [inline culinary-canvas/capstone-project/frontend/my_capstone_project/src/Notification.jsx]: do they see the contents of the notification without clicking on it?

File: culinary-canvas/capstone-project/frontend/my_capstone_project/src/Notification.jsx [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+    const markAsRead = async(notificationId) => {
+        try{
+            await fetch(`${backendAddress}/notifications/read`, {
+                method: 'POST',
+                headers: {'Content-Type' : 'application/json'},
+                body: JSON.stringify({notificationId}),
+                credentials: 'include'
+            });
+            setNotifications(notifications.map(n => n.id === notificationId ? {...n, read: true} : n))
```
Comment (SamirMGH) [inline culinary-canvas/capstone-project/frontend/my_capstone_project/src/Notification.jsx]: Small suggestion: Instead of using map, you can use findIndex to see if the specific notification exists and modify just that one element's 'read' to true.

Edit: Actually never mind, since it's not a sorted search it won't yield much of an improvement although it'd avoid the small cost of making a copy here

File: culinary-canvas/capstone-project/frontend/my_capstone_project/src/UserProfile.jsx [hunk 9]
```diff
@@ -8,13 +8,65 @@ import SavedRecipes from "./SavedRecipes";
...
+        }catch(error){
+          console.error('Error fetching health scores: ', error)
+        }
+      };
+      fetchHealthScores();
+      socket.on('connect', () => {
+        socket.emit('suscribe', user)
+    });
+
+      socket.on('healthScoresUpdated', (updatedHealthScores) => {
```
Comment (SamirMGH) [inline culinary-canvas/capstone-project/frontend/my_capstone_project/src/UserProfile.jsx:48]: Indenting is a bit off here. If it's possible to use an auto-formatter, definitely give it a shot! It helps with code logic readability.

**Comment by OluwadaraO** [comment]
https://www.loom.com/share/9c45bef5c03d4354ab6ff4a9d87ec203

This is the demo of the push notifications.

**Comment by OluwadaraO** [comment]
Screenshot 2024-07-25 at 10 40 37 PM
Notifications being sent to user when they're not on that tab.
